magda-indexer/src/test/scala/au/csiro/data61/magda/indexer/helpers/StreamControllerTest.scala

```diff
@@ -0,0 +1,215 @@
+        })
+        .runWith(Sink.fold(Future(SearchIndexer.IndexResult(0, Seq()))) {
+          case (combinedResultFuture, (thisResultIdentifier, thisResultFuture)) =>
+            combinedResultFuture.flatMap { combinedResult =>
+              thisResultFuture.map { _ =>
+                combinedResult.copy(successes = combinedResult.successes + 1)
+              }.recover {
+                case _: Throwable =>
+                  combinedResult.copy(failures = combinedResult.failures :+ thisResultIdentifier)
+              }
+            }
+        })
+
+      indexResults.flatMap(identity)
+    }
+
+    private def index(dataSet: DataSet, promise: Promise[Unit] = Promise[Unit]): Future[Unit] = {
+      dataSetCount.incrementAndGet()
+      buffer.append(promise)
+
+      if (dataSetCount.get() == dataSets.size) {
+        // Simulate non-full batch processing
+        buffer.toList.foreach(promise => promise.success())
+        buffer.clear()
+      }
+      else if (dataSetCount.get() % batchProcessingSize == 0) {
+        // Simulate batch processing
+        buffer.toList.foreach(promise => promise.success())
+        buffer.clear()
+      }
+
+      promise.future
+    }
+
+    override def delete(identifiers: Seq[String]): Future[Unit] = {
+      throw new Exception("delete() not implemented")
+    }
+
+    override def snapshot(): Future[Unit] = {
+      throw new Exception("snapshot() not implemented")
+    }
+
+    override def ready: Future[Unit] = {
+      throw new Exception("ready() not implemented")
+    }
+
+    override def trim(before: OffsetDateTime): Future[Unit] = {
+      throw new Exception("trim() not implemented")
+    }
+
+    override def isEmpty(index: Indices.Index): Future[Boolean] = {
+      throw new Exception("isEmpty() not implemented")
+    }
+  }
+
+  private def createDataSets(num: Int): Seq[DataSet] = {
+    val range = 1 to num
+    range.foldLeft(Seq[DataSet]()){
+      case (acc, e) =>
+        acc ++ Seq(
+          DataSet(identifier = s"d$e", catalog = Some("c"), quality = 1.0D, score = Some(1.0F)))
+    }
+  }
+
+  private val bufferSize = 8
+  private def run(dataSetNum: Int): Future[Assertion] = {
+    dataSets = createDataSets(dataSetNum)
+    val mockRegistryInterface = new MockRegistryInterface()
+    sc = new StreamController(mockRegistryInterface, bufferSize)
+    val source = sc.getSource
+    val mockIndexer = new MockIndexer(2)
+    val indexResultF: Future[IndexResult] = mockIndexer.index(source)
+
+    sc.start()
+
+    indexResultF.map(indexResult =>
+      indexResult shouldEqual IndexResult(dataSets.size, List()))
+  }
+
+  "The stream controller" should "support the indexer stream 1" in {
+    val numOfDataSets = 10 * bufferSize - 1
+    run(numOfDataSets)
+  }
+
+  "The stream controller" should "support the indexer stream 2" in {
+    val numOfDataSets = 10 * bufferSize
+    run(numOfDataSets)
+  }
+
+  "The stream controller" should "support the indexer stream 3" in {
```

Comment:
Can we get some more descriptive test descriptions - what is this test doing that the other ones aren't? If this test is the only one that fails I'm going to have no idea what it specifically does that's failing.

Reply:
They are testing the control logic at edge cases. Given a source buffer size, it should work with any number of datasets. It is easy to make logical errors for those dataset numbers. How about changing them like this:
"The stream controller" should "support the indexer stream when the number of datasets is xxx".

Reply:
👍

Reply:
Please note that StreamControllerTest was written in the early stage when MockIndexer still tries to mock the old index function. However, the control logic still holds.